lib/posthog/feature_flags.rb

```diff
@@ -42,96 +77,288 @@ def is_feature_enabled(key, distinct_id, default_result = false, groups = {})
+        return override_value.class == value.class && override_value < value
+      elsif operator == 'lte'
+        return override_value.class == value.class && override_value <= value
       else
-        res = get_feature_variants(distinct_id, groups)
-        return res[key] ? true : default_result
+        return false
       end
 
-      return false
     end
 
-    def is_simple_flag_enabled(key, distinct_id, rollout_percentage)
-      hash = Digest::SHA1.hexdigest "#{key}.#{distinct_id}"
-      return(
-        (Integer(hash[0..14], 16).to_f / 0xfffffffffffffff) <=
-          (rollout_percentage / 100)
-      )
+    def self.is_valid_regex(regex)
+      begin
+        Regexp.new(regex)
+        return true
+      rescue RegexpError
+        return false
+      end
     end
 
-    def load_feature_flags(force_reload = false)
-      if @loaded_flags_successfully_once.false? || force_reload
-        _load_feature_flags
+    private
+
+    def _compute_flag_locally(flag, distinct_id, groups = {}, person_properties = {}, group_properties = {})
+      if flag['ensure_experience_continuity']
+        raise InconclusiveMatchError.new("Flag has experience continuity enabled")
+      end
+
+      if !flag['active']
+        return false
+      end
+
+      flag_filters = flag['filters'] || {}
+      aggregation_group_type_index = flag_filters['aggregation_group_type_index']
+      if !aggregation_group_type_index.nil?
+        group_name = @group_type_mapping[aggregation_group_type_index.to_s]
+
+        if group_name.nil?
+          logger.warn "[FEATURE FLAGS] Unknown group type index #{aggregation_group_type_index} for feature flag #{flag['key']}"
+          # failover to `/decide/`
+          raise InconclusiveMatchError.new("Flag has unknown group type index")
+        end
+
+        if !groups.has_key?(group_name)
+          # Group flags are never enabled if appropriate `groups` aren't passed in
+          # don't failover to `/decide/`, since response will be the same
+          logger.warn "[FEATURE FLAGS] Can't compute group feature flag: #{flag['key']} without group names passed in"
+          return false
+        end
+
+        focused_group_properties = group_properties[group_name]
+        return match_feature_flag_properties(flag, groups[group_name], focused_group_properties)
+      else
+        return match_feature_flag_properties(flag, distinct_id, person_properties)
       end
+
     end
 
-    def get_feature_variants(distinct_id, groups = nil)
-      groups = {} unless groups.is_a?(Hash)
+    def match_feature_flag_properties(flag, distinct_id, properties)
+      flag_filters = flag['filters'] || {}
+      flag_conditions = flag_filters['groups'] || []
+      is_inconclusive = false
+      result = nil
 
-      request_data = {
-            "distinct_id": distinct_id,
-            "personal_api_key": @personal_api_key,
-            "groups": groups
-      }
+      flag_conditions.each do |condition|
```

Comment:
Suggestion: make this code functional. There's a lot going on her with `InconclusiveMatchError` and what not to make providing a straight-forward solution here tricky but it is also likely a code smell.